src/models/spaces/populations/mouse-model/hawks.ts

```diff
@@ -1,16 +1,66 @@
-import { BasicAnimal, Species, Trait, Agent } from "populations.js";
+import { BasicAnimal, Species, Trait, Agent, AgentDistance } from "populations.js";
+
+let nextZIndex = 2;
 
 class Hawk extends BasicAnimal {
 
+  private _circleCount = 0;
+  private _baseZIndex = nextZIndex;
+
   constructor(args: any) {
     super(args);
     this.label = "Hawk";
     this._viewLayer = 3;
+    (this as any).zIndex(this._baseZIndex);
+    nextZIndex++;
   }
 
   public isInteractive() {      // FIXME: Would be nice to set this better in populations
     return false;
   }
+
+  public makeNewborn() {
+    super.makeNewborn();
+    this._circleCount = Math.random() * 2;
+  }
+
+  public wander() {
+    let newHeight = this.get("height") + 1;
+    newHeight = Math.min(newHeight, 10);
+    this.set("height", newHeight);
+    // a lower hawk is always below a higher one, and at the same level one is always higher
+    (this as any).zIndex(this._baseZIndex + (newHeight * 10));
+    this.set("current behavior", BasicAnimal.BEHAVIOR.WANDERING);
+    const speed = this.get("speed") - 3;
+    const delta = Math.sqrt(10 / (1 + (9 * Math.pow(Math.cos(this._circleCount), 2)))) * Math.cos(this._circleCount);
+    let newDir = this.get("direction") + (delta / 25);
+    this.set("direction", newDir);
+    // if we are nearing the edges, turn harder
+    const loc = this.getLocation();
+    if (loc.x < 120 || loc.x > (this as any).environment.width - 120
+        || loc.y < 120 || loc.y > (this as any).environment.height - 120) {
+      const dx = speed * 20 * Math.cos(newDir);
+      const dy = speed * 20 * Math.sin(newDir);
+      if ((this as any).environment.crossesBarrier(loc, {
+        x: loc.x + dx,
+        y: loc.y + dy
+      })) {
+        newDir = this.get("direction") + (delta / 15);
+        this.set("direction", newDir);
+      }
+    }
+    this._circleCount += 0.01;
+    return this.move(speed);
+  }
+
```

Comment:
This movement handling all seems good.  Obviously we still have the jitters due to eating, but there is only so much we can do without radically changing what happens when a hawk eats a mouse (a particle effect, pop-up text, or some small animation effect to convey that a mouse has been consumed would be cool, but definitely not on the table for this round).  I actually like when the hawk movement brings the hawks slightly outside the bounds of the sim.  I think this looks neat.

Reply:
Yes, I actually would have liked to explore having the bounds of the model being larger than the bounds of the view, but that's tricky to do, since people wanted students to be able to see every mouse, to avoid the case where a mouse of a different color was outside the field of view and causing confusion with graphs or breeding.

The way to solve that would be to have the bigger model, and to have barriers to stop the mice from exiting the visible view. But right now we can't have different barriers for different species, so the hawks would be bound by it too.